mosaicml/composer
# GHA Docker build flow for PR's

# What does this PR do?

Starting to move Docker builds to GHA.  In this PR we:
* Create the initial reusable Docker build workflow (`docker-build.yaml`)
   * Takes an image config YAML as input
   * Add's supporting script (`gen_docker_matrix.yaml`) to translate an image config YAML and translate it to a GHA matrix 
   * Builds and optionally pushes images to a specified Docker repo
   * Requires repo DockerHub secrets
   * Reusable workflow still needs some work to be completely generic, cleanup will be handled in subsequent PR's as top-level pipelines are developed.
* Create the top level `PR Docker` workflow (`pr-docker.yaml`)
   * Runs in PR's whenever repo Docker or Docker build CI files are touched
* Updates image config generation script, adds `IMAGE_NAME` parameter
* Upgrades `flash-attn` to 0.2.8
* Testing for this PR:
   * Docker builds pass in GHA
   * Confirm images are pushed successfully to staging repo
   * Confirm image cache's are generated based on `IMAGE_NAME`
   * Confirm image cache works by re-triggering workflow after initial successful build, confirm speed-up
   * Manually verify `pytest` test suite works on Torch 1.13 images
* Outside the scope of this PR:
   * We don't confirm that the `pytest` test suite works with all new images, this will be tested in a follow-up PR that merges the Docker build with the `pytest` testing workflows
   * Releasing Docker images to prod DockerHub repo's: `mosaicml/pytorch`, `mosaicml/pytorch_vision`, `mosaicml/composer`

# What issue(s) does this change relate to?

Fixes [CO-1064](https://mosaicml.atlassian.net/browse/CO-1064)
Fixes [CO-1606](https://mosaicml.atlassian.net/browse/CO-1606)

# To Do's Before Submitting
- [x] Ask IT to create a `mosaicml/ci-staging` repo, migrate off of `mosaicml/jenkins-staging`
- [x] Disable Jenkins Docker CI flows

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))


[CO-1064]: https://mosaicml.atlassian.net/browse/CO-1064?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
